### ISSUE TITLE
Polyglot Language Model

### ISSUE THREAD

**Issue by eisenjulian** [n-waves/ulmfit-multilingual#15]
- [x] Pretrain a language model with for two wikipedias (such as en and de) with a single vocabulary from sentencepiece
- [x] Prepare RCV data
- [x] Prepare CLS data
- [x] Test zero shot classification accuracy on XNLI/CLS/RCV. The hope is that with none or only a few training samples from a target language and full training set from a source language we can learn a good classifier
- [ ] Test alignment loss term from Europarl parallel corpus (such as what they do in XNLI paper)

**Comment by dust0x** [comment]
Hi Julian, any progress with this? If you need some extra help with the experiments, I can contribute.

**Comment by eisenjulian** [comment]
Hey @dust0x thanks for the offer. that would be great. So far I have a working [branch](https://github.com/n-waves/ulmfit-multilingual/compare/datasets) to format RCV and CLS multilanguage classification datasets and [another one](https://github.com/n-waves/ulmfit-multilingual/compare/polyglot-lm) to create a dataset alternating tokens from two languages and train the bilingual model. 

I've run into a bunch of memory issues and I've only completed training a small model so far :(. I am currently training a bigger model (10 million lines of training data) but it looks like it will take +24 hours to run 10 epochs on a P100. I am looking to add a callback to save checkpoints along the way as well as avoid retokenizing every time (about 40 minutes on 8 vcpu 56 gigs mem) since this is very time-consuming at the moment.

Currently what I am running is 
```sh
python ulmfit/merge_langs.py '["~/data/wiki/en-100", "~/data/wiki/de-100"]' '["en", "de"]' ~/data/wiki/en-de-100 10
python -m ulmfit.pretrain_lm ~/data/wiki/en-de-100 en-de --subword --max_vocab=30000 --noqrnn
```
If you want to give it a go as well, maybe you will have more luck. Once the model is trained, the evaluation part should be the same as usual, just using the sentencepiece model.

**Comment by eisenjulian** [comment]
I added caching of the tokenized dataset and exporting model after every epoch to the branch. I just finished running 6 epochs (and counting) and will run some validations. If they look well I can upload the model to a shared CDN and we can start evaluating on the some of the datasets.

@dust0x where you able to train the English sentencepiece model, it would be great to 1) know how the two models compare and 2) see which hyperparams worked best in that setting.

**Comment by dust0x** [comment]
Sorry, I haven't trained an end-to-end model with sentencepiece. I currently don't have sufficient compute for it, I can only run small tests. I might be able to get a GTX 1060 but that's really the best I'll be able to do for now.

**Comment by eisenjulian** [comment]
@PiotrCzapla @dust0x  while running some sanity checks I noticed a bug [here](https://github.com/n-waves/fastai/blob/ulmfit_multilingual/fastai/text/data.py#L166) the fastai branch that we're using. It's basically ignoring the tokenizer when loading the dataset. It seems to have been fixed in this [commit](https://github.com/fastai/fastai/commit/55ceefdda35a2646dd0bb8f647e982aba5c99b74) upstream, do you think it's safe to cherry pick or upgrade? Or should I just fix that line?

**Comment by sebastianruder** [comment]
Slightly off-topic, but we might want to consider making this repo private and only selectively giving access to avoid other people raising issues while it's still work in progress. What do you think?

**Comment by PiotrCzapla** [comment]
@eisenjulian  I was sure I fixed it and pushed to the n-waves/fastai repo but it seems I haven't.  Thank you for noticing that. Yesterday I've updated the refactoring branch to use the newest fastai. I'm working on the classifier code to use multiple different encodings once this is done I  will push the final version to `refactoring` branch and we can update n-waves/fastai . 

@sebastianruder  I would prefer to keep the repo public so that others that want to do classification can benefit, and I'm going to fix the bugs if they report them.

**Comment by sebastianruder** [comment]
Ok. 👍

**Comment by PiotrCzapla** [comment]
@eisenjulian I'm glad you liked the refactoring branch. Re. MLDoc I'm trying to get access to that dataset as well, I'm wondering how you managed to get it as the routers website is down,  can we talk on priv. twitter.com/piotrczapla or forums.fastai @piotr.czapla?

I've noticed the paper, some time ago but I haven't had a moment to act on it. I would love to explore the concept but first I'd like to test ulmfit trained on single languages on MLDoc to get baseline.

**Comment by eisenjulian** [comment]
Sure, I have run MLDoc on classic ULMFit in a few languages in the past (maybe 6 months ago) I can dig out those results, they were good AFAIR, but we didn't have a fair benchmark to compare it against, since most papers were using zero-shot learning approaches, and beating that was quite easy.

Reach out on twitter to me at @eisenjulian I checked but I think your DMs are closed. I can dig out those results, and I also have exported data as csvs in closed GCP bucket I can share privately, since the raw RCV data is licensed :(

**Comment by eisenjulian** [comment]
Will close this issue since this line of work didn't yield good results and we are not actively working on it anymore
